yourbench/pipeline/chunking.py

```diff
@@ -0,0 +1,608 @@
+# =============================================================================
+# chunking.py
+# =============================================================================
+"""
+@module chunking
+@author @sumukshashidhar
+
+This module implements the "Semantic Chunking" stage of the YourBench pipeline.
+It takes ingested and optionally summarized documents, partitions them into
+multiple coherent segments (single-hop chunks), and optionally creates multi-hop
+chunks by sampling and concatenating various single-hop segments.
+
+Preserves semantic relationships among sentences by leveraging embeddings from a
+transformer-based model (e.g., "intfloat/multilingual-e5-large-instruct"). This
+stage helps downstream question generation avoid handling entire long documents
+at once, improving coverage and reducing the risk of overlooking important but
+less prominent content.
+
+Usage:
+------
+Typically, you do not call this module directly. Instead, the `handler.py`
+automatically invokes `run(config)` if the corresponding pipeline setting
+(`pipeline.chunking.run`) is enabled.
+
+The `run(config)` function:
+1. Loads a dataset specified by the pipeline configuration.
+2. Splits each document into single-hop chunks, guided by user-defined token
+   length constraints (`l_min_tokens`, `l_max_tokens`) and a similarity threshold
+   (`tau_threshold`).
+3. Creates multi-hop chunks by sampling subsets of single-hop chunks and
+   concatenating them.
+4. Computes optional readability and perplexity metrics for each chunk if debug
+   mode is enabled.
+5. Saves the dataset containing new columns:
+   - "chunks" (list of single-hop segments)
+   - "multihop_chunks" (list of multi-hop segment groups)
+   - "chunk_info_metrics" (various statistics)
+   - "chunking_model" (the model used for embeddings).
+
+Error Handling and Logging:
+---------------------------
+- All warnings, errors, and debugging information are logged to both the console
+  and a dedicated log file at `logs/chunking.log`.
+- If any critical errors occur while loading or processing data, the process
+  logs the exception and attempts a graceful exit without crashing the entire
+  pipeline.
+
+"""
+
+import itertools
+import os
+import random
+import re
+from typing import Any, Dict, List, Optional
+
+import evaluate
+import matplotlib.pyplot as plt
+import torch
+import torch.nn.functional as F
+from loguru import logger
+from transformers import AutoModel, AutoTokenizer
+
+from yourbench.utils.dataset_engine import (
+    save_dataset,
+    smart_get_output_dataset_name,
+    smart_get_output_subset,
+    smart_get_source_dataset_name,
+    smart_get_source_subset,
+    smart_load_dataset,
+)
+
+
+# === Stage-Specific Logger Configuration ===
+os.makedirs("logs", exist_ok=True)
+logger.add("logs/chunking.log", level="DEBUG", rotation="10 MB", enqueue=True, backtrace=True, diagnose=True)
```

Comment:
in the future, we'll likely want to add a UID based on the time to avoid overwriting/appending to previously existing file